common/flatpak-transaction.c

```diff
@@ -4062,6 +4092,272 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
     }
 }
 
+static gboolean
+installed_ref_ref_equal (gconstpointer a,
+                         gconstpointer b)
+{
+  FlatpakInstalledRef *iref = (FlatpakInstalledRef *)a;
+  const char *needle_ref = (const char *)b;
+  g_autofree char *full_ref = flatpak_ref_format_ref (FLATPAK_REF (iref));
+  return g_strcmp0 (full_ref, needle_ref) == 0;
+}
+
+static GPtrArray *
+find_related_from_deploy (FlatpakTransaction  *self,
+                          const char          *ref,
+                          char               **out_remote)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr(GBytes) metadata_bytes = NULL;
+  g_autoptr(GKeyFile) metakey = NULL;
+  g_autofree char *checksum = NULL;
+  g_autofree char *remote = NULL;
+  g_autoptr(GError) related_error = NULL;
+  g_autoptr(GPtrArray) related = NULL;
+
+  metadata_bytes = load_deployed_metadata (self, ref, &checksum, &remote);
+  if (metadata_bytes == NULL)
+    return NULL;
+
+  if (out_remote)
+    *out_remote = g_strdup (remote);
+
+  metakey = g_key_file_new ();
+  if (!g_key_file_load_from_bytes (metakey, metadata_bytes, G_KEY_FILE_NONE, NULL))
+    {
+      g_message ("Warning: Failed to parse metadata for %s\n", ref);
+      return NULL;
+    }
+
+  related = flatpak_dir_find_local_related_for_metadata (priv->dir, ref,
+                                                         checksum, remote, metakey,
+                                                         NULL, &related_error);
+  if (related_error != NULL)
+    g_message (_("Warning: Problem looking for related refs: %s"), related_error->message);
+
+  return g_steal_pointer (&related);
+}
+
+static gboolean
+prune_maybe_unused_list (FlatpakTransaction  *self,
+                         GHashTable          *maybe_unused_runtimes,
+                         GPtrArray           *to_be_excluded,
+                         GCancellable        *cancellable,
+                         GError             **error)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  GVariantBuilder builder;
+  g_autoptr(GVariant) list_unused_options = NULL;
+  g_autoptr(GPtrArray) unused_refs = NULL;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sv}"));
+  if (to_be_excluded->len > 0)
+    {
+      g_ptr_array_add (to_be_excluded, NULL);
+      g_variant_builder_add (&builder, "{s@v}", "exclude-refs",
+                             g_variant_new_variant (g_variant_new_strv ((const char * const *) to_be_excluded->pdata, -1)));
+    }
+  g_variant_builder_add (&builder, "{s@v}", "filter-by-eol",
+                         g_variant_new_variant (g_variant_new_boolean (TRUE)));
+  list_unused_options = g_variant_ref_sink (g_variant_builder_end (&builder));
+
+  unused_refs = flatpak_installation_list_unused_refs_with_options (priv->installation,
```

Comment:
I would prefer if this didn't call the highlevel installation API directly, but rather a helper function that just returns the refs. Because flatpak_installation_list_unused_refs_with_options() essentially create the list of refs, and then calls get_ref() on each of them, which reads a bunch of data like the deploy files. Then beliow we throw all that away and format the InstalleRef as a string which we compare with another string.

Maybe you can make a version of flatpak_installation_list_unused_refs_with_options() like flatpak_dir_list_unused_refs_with_options() which returns a list of ref strings, and call that from the flatpak_installation_ API combined with get_ref(), and from the transaction code here.

Reply:
You can make the flatpak_dir version not use the gvariant for options, because it is internal APIs we can extend however we want later.

Reply:
Well we are reading the deploy data either way because we use it to find the runtime and other info. But reading it once is better than reading it twice, and in your scheme we could also avoid the cost of the `flatpak_dir_read_latest()` and `flatpak_dir_get_remote_collection_id()` calls. So I guess it's worth the rearrangement.

Reply:
Added a commit for this